cmd/zoekt-sourcegraph-indexserver/sg.go

```diff
@@ -82,19 +96,29 @@ func (s *sourcegraphClient) List(ctx context.Context, indexed []uint32) (*Source
 		// We ask the frontend to get index options in batches.
 		for repos := range batched(repos, batchSize) {
 			start := time.Now()
-			opts, err := s.GetIndexOptions(repos...)
+			opts, fingerprint, err := s.getIndexOptions(lastFingerprint, repos...)
 			if err != nil {
+				// Call failed, restore old fingerprint for next call to List.
+				first = false
+				s.configFingerprint.Store(lastFingerprint)
```

Comment:
This feels like iterate should return an error `err`. If `err == nil` we update the fingerprint, otherwise it stays as is. I think that would simplify the logic around `first`, too. WDYT?

Reply:
iterate ignoring errors is copying the older behaviour. I agree though, it feels like if one of these calls fails there is no point continueing through the batch. I'll follow-up with a fix here since I want to retain old behaviour here and more properly think it through.

I'm not sure it would necessarily simplify logic around first, since that is needed to only get the cursor of the first call. This is the only bit which isn't really opaque and requires understanding the temporal stuff around the fingerprint. Maybe there is a way to make that clearer?

Reply:
@stefanhengl I just tried updating this, and not sure it made the code better. Can you sketch out the idea you have / show a diff? Here is the minimal diff implementing your idea as I interpretted it

```patch
diff --git a/cmd/zoekt-sourcegraph-indexserver/sg.go b/cmd/zoekt-sourcegraph-indexserver/sg.go
index f070760..6f76ebe 100644
--- a/cmd/zoekt-sourcegraph-indexserver/sg.go
+++ b/cmd/zoekt-sourcegraph-indexserver/sg.go
@@ -99,13 +99,12 @@ func (s *sourcegraphClient) List(ctx context.Context, indexed []uint32) (*Source
                        opts, fingerprint, err := s.getIndexOptions(lastFingerprint, repos...)
                        if err != nil {
                                // Call failed, restore old fingerprint for next call to List.
-                               first = false
                                s.configFingerprint.Store(lastFingerprint)

                                metricResolveRevisionDuration.WithLabelValues("false").Observe(time.Since(start).Seconds())
                                tr.LazyPrintf("failed fetching options batch: %v", err)
                                tr.SetError()
-                               continue
+                               return
                        }

                        if first {

```